src/environment/Settings.cpp

```diff
@@ -0,0 +1,116 @@
+/*
+ *   AviTab - Aviator's Virtual Tablet
+ *   Copyright (C) 2020 Folke Will <folko@solhost.org>
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU Affero General Public License as published by
+ *   the Free Software Foundation, either version 3 of the License, or
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU Affero General Public License for more details.
+ *
+ *   You should have received a copy of the GNU Affero General Public License
+ *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <fstream>
+#include "src/platform/Platform.h"
+#include "Settings.h"
+#include "src/Logger.h"
+#include <nlohmann/json.hpp>
+
+using json = nlohmann::json;
+
+namespace avitab {
+
+Settings::Settings(const std::string &settingsFile)
+:   filePath(settingsFile)
+{
+    database = std::make_shared<json>();
+    init();
+    load();
+    upgrade();
+}
+
+Settings::~Settings()
+{
+    save();
+}
+
+template<>
+bool Settings::getGeneralSetting(const std::string &id) {
+    json::json_pointer jp("/general/"+id);
+    bool b = database->value(jp,false);
+    return b;
+}
+
+template<>
+int Settings::getGeneralSetting(const std::string &id) {
+    json::json_pointer jp("/general/"+id);
+    int b = database->value(jp,0);
+    return b;
+}
+
+template<>
+void Settings::setGeneralSetting(const std::string &id, const bool value) {
+    json::json_pointer jp("/general/"+id);
+    (*database)[jp] = value;
+}
+
+template<>
+bool Settings::getOverlaySetting(const std::string &id) {
+    json::json_pointer jp("/overlay/"+id);
+    bool b = database->value(jp,false);
+    return b;
+}
+
+template<>
+void Settings::setOverlaySetting(const std::string &id, const bool value) {
+    json::json_pointer jp("/overlay/"+id);
+    (*database)[jp] = value;
+}
+
+void Settings::init() {
+    // full init not required, just defaults that aren't zero/false/empty
+    *database = { { "general", { { "prefs_version", 1 }, { "show_fps", true } } }, { "overlay", { { "my_aircraft", true } } } };
+}
+
+void Settings::load() {
+    json filedata;
+    try {
+        fs::ifstream fin(fs::u8path(filePath));
+        fin >> filedata;
+    } catch (const std::exception &e) {
+        LOG_WARN("Could not load user settings from %s, using defaults", filePath.c_str());
+    }
+    for (const auto &i : filedata.items()) {
+        (*database)[i.key()] = i.value();
+    }
+}
+
+void Settings::upgrade() {
+    // handle older json databases which have subsequently been updated
+#if 0 // example idea for this code
+    if (getGeneralSetting<int>("prefs_version") == 1) {

```

Comment:
Hm.. another approach that doesn't require upgrades would be moving the default value to the caller, i.e. `getSettings("mySettings", true)` to make the default value `true` if the value doesn't exist. That way, both the `init` and `upgrade` functions wouldn't be necessary in my opinion. Just an idea, I'll let you decide.